tests/unit/test_data.py

```diff
@@ -160,6 +173,53 @@ def test_normalize(self, data_type, weights):
             normalized_with_zeros.weights, jnp.nan_to_num(expected_weights)
         )
 
+    def test_jaxtyping_compatibility(self, data_type):
+        """
+        Test `Data` compatibility with jaxtyping annotations.
+
+        Checks the following cases:
+            - Correct shape and data type,
+            - Correct shape and incorrect data type,
+            - Incorrect shape and correct data type.
+        """
+        float_data_type = eqx.tree_at(
+            lambda x: x.args,
+            data_type,
+            replace=jtu.tree_map(lambda y: jnp.astype(y, jnp.float32), data_type.args),
+        )
+        int_data_type = eqx.tree_at(
+            lambda x: x.args,
+            data_type,
+            replace=jtu.tree_map(lambda y: jnp.astype(y, jnp.int32), data_type.args),
+        )
+        float_data = float_data_type()
+        int_data = int_data_type()
+
+        @jaxtyped(typechecker=beartype)
+        def good_shape_check(x: Shaped[coreax.data.Data, "3 1"]):
+            del x
+
+        @jaxtyped(typechecker=beartype)
+        def float_dtype_check(x: Float[coreax.data.Data, "..."]):
+            del x
+
+        @jaxtyped(typechecker=beartype)
+        def bad_shape_check(x: Shaped[coreax.data.Data, "3"]):
+            del x
+
+        float_check_context = pytest.raises(jaxtyping.TypeCheckError)
+        bad_shape_check_context = pytest.raises(jaxtyping.TypeCheckError)
+        for data in [int_data, float_data]:
+            good_shape_check(data)
+            with bad_shape_check_context:
+                bad_shape_check(data)
+            # We don't expect an error to be raised in the `float_data`` case, on a call
+            # to float_dtype_check(...)
+            if data is float_data:
+                float_check_context = nullcontext()
+            with float_check_context:
+                float_dtype_check(data)
```

Comment:
For this entire section, rather than defining functions to check with `@jaxtyped` and checking if they error, it'd be much simpler to just use `assert beartype.door.is_bearable(data, type_hint)` (or `assert not`, as required) to check whether or not a type hint matches.

Reply:
I like this idea! Not sure why I made this so complicated first time round.